Create bundle.md (#341)
sigstore/docs: content/en/about/bundle.md

# Sigstore Bundle

October 31, 2024

Version 0.3.2

This document describes the data structure for storing Sigstore signatures generated by tooling
working in the context of the Sigstore Public Instance. It includes `json` examples of
serialized bundles of the current bundle format version. It may exclude descriptions of parameters
that continue to exist for compatibility reasons or for private use cases. For a full description of
the format, the formal schema and information about language library support see
[sigstore/protobuf-specs](https://github.com/sigstore/protobuf-specs).

## Bundle

A Sigstore bundle is everything required to verify a signature on an artifact. This
is satisfied by the **Verification Material** and signature **Content**.

### Verification Material

> This is key materical used to verify signatures along with supporting metadata like transparency log entries and timestamps. Bundles must include at least one transpareny log's signed entry timestamp or an rfc3161 timestamp to provide proof of signing time.

Reviewer: ```suggestion
This is key material used to verify signatures along with supporting metadata like transparency log entries and timestamps. Bundles must include at least one transparency log's signed entry timestamp or an [RFC3161](https://www.ietf.org/rfc/rfc3161.txt) timestamp to provide proof of signing time.
```

#### Key Material

A single X.509 leaf certificate conveying the signing key and containing [extensions](https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md) 
for identities consumed at verification time. This is the recommended `"verificationMaterial"` type

> for use with the public Sigstore infrastructure.

Reviewer: Can we mention `PublicKeyIdentifier` as another option? Using a self-managed key and recording the signature in the public Rekor instance is a supported workflow.

Author: done

```json
"verificationMaterial": {
  "certificate": {
    "rawBytes": "<Base64(DER bytes)>"
  }
}
```

#### Transparency Log Entries

One or more transparency logs entries to provide proof of inclusion in a public log and optionally a timestamp to

> validate signing occurred at a valid point in time.

Reviewer: I'd define what we mean by "valid point in time", as in a user was in possession of a signing key and used it to produce a signature, and that signing key was certified by a CA, and this is recorded while the certificate is valid in the log.

Author: done

```json
"verificationMaterial": {
  "tlogEntries": [
    {
      "logIndex": "<GLOBAL_LOG_INDEX>",
      "logId": {
        "keyId": "<KEY_ID>",
      },
      "kindVersion": {
        "kind": "(hashrekord | dsse)",
        "version": "<VERSION>"
      },
      "integratedTime": "<UNIX_TIMESTAMP>",
      "inclusionPromise": {
        "signedEntryTimestamp": "<Base64(SET bytes)>"
      },
      "inclusionProof": {
        "logIndex": "<TREE_INDEX>",
        "rootHash": "<Base64(HASH)>",
        "treeSize": "<TREE_SIZE>",
        "hashes": [
          "<Base64(HASH)>",
          "<Base64(HASH)>",
          "<Base64(HASH)>"
        ],
        "checkpoint": {
          "envelope": "<TLOG CHECKPOINT SIGNED NOTE>"
        }
      },
      "canonicalizedBody": "<Base64(JSON)>"
    },
  ]
}
```

#### Timestamp

> Zero or more [RFC3161](https://www.ietf.org/rfc/rfc3161.txt) timestamps to validate signing occurred at a valid point in time

Reviewer: Same here to clarify what "valid" means

Author: done

```json
"verificationMaterial": {
  "timestampVerificationData": {
    "rfc3161Timestamps": [
      {
        "signedTimestamp": "Base64(RFC3161 TIMESTAMP)"
      },
    ]
  }
}
```

### Content

This is the signature data for which the Verification Material is defined over. It must be one of

> **Message Signature** over an artifact hash or a **DSSE** envelope for for attestations.

Reviewer: ```suggestion
**Message Signature** over an artifact hash or a **DSSE** envelope for attestations.
```

#### Message Signature

This is a computed signature over a message (typically an arifact hash). It may contain a
`message_digest` for informational purposes, but it must be provided or computed from a provided
artifact at verification time.

```json
"messageSignature": {
  "messageDigest": {
    "algorithm": "<HASH ALGORITHM>",
    "digest": "<Base64(DIGEST)>"
  },
  "signature": "<Base64(SIGNATURE)>"
}
```

#### DSSE
 

> A DSSE envelope can contain arbitrary payloads. Currently clients only process the

Reviewer: ```suggestion
A DSSE envelope can contain arbitrary payloads. Currently Sigstore clients only process the
```

#### Key Material

##### X.509 certificate

A single X.509 leaf certificate conveying the signing key and containing [extensions](https://github.com/sigstore/fulcio/blob/main/docs/oid-info.md)
for identities consumed at verification time. This is the recommended `"verificationMaterial"` type
for use with the public Sigstore infrastructure.

```json
"verificationMaterial": {
  "certificate": {
    "rawBytes": "<Base64(DER bytes)>"
  }
}
```

##### Public Key Identifier

A hint to identify an out of band delivered key to verify a signature. Like traditional PKI key distribution
the format of the hint must be agreed upon out of band by the signer and the verifiers. The key itself is not embedded in the Sigstore bundle.

```json
"verificationMaterial": {
  "publicKeyIdentifier": {
    "hint": "<HINT>"
  }
}
```

#### Transparency Log Entries

> One or more transparency logs entries to provide proof of inclusion in a public log and optionally a timestamp to

Reviewer 2: Minor: a transparency log entry isn't required for a bundle, should we say "Zero or more..." like we do below for Timestamps?

We already say above:
> bundles must include at least one transparency log's signed entry timestamp or an RFC3161 timestamp

Author: Oh I always thought a log entry was required but you don't have to trust the logs clock (so you can ignore the SET and use the rfc3161 timestamp). Maybe that's only true for the public usecase?

Reviewer: Log entries are not required - https://github.com/sigstore/protobuf-specs/blob/main/protos/sigstore_bundle.proto#L105

I think we should word it that it is expected that bundles contain log entries for public consumption.

Author: updated this section a bit
